[WIP] Implements of in-app EXIF and location anonymization

ilgazer opened the pull request:
## Title (required)

Fixes #1686 Implementation of in-app EXIF modification
Fixes #212 Remove from EXIF: copyright field (name)
Fixes #181 Remove from EXIF: geo info if too close to home
Fixes #70 Option to "anonymize" pictures 

## Description (required)

There are multiple issues about privacy concerns location and EXIF-related privacy concerns. These commits implement two different settings for removing certain EXIF rags from the picture and for reducing the accuracy or completely eliminating geolocation data from the picture.

The bulk of the anonymization logic is implemented in FileProcessor to prevent code reuse. The existing coordinate extraction logic in MultipleShareActivity is also replaced by a FileProcessor to further reduce code reuse.

Add method anonymizeCoord() that reduces the resolution of GPS coordinates. Think of each coordinate as a pixel. This function reduces the resolution of "the world" to one coordinate(pixel) for every x kilometres (x is the variable prefLocationAccuracy). This approach was chosen over changing the coordinates by a random amount because with that approach; if a user uploads sufficiently many pictures from the same location, statistical analysis can reveal the real location of the user. 

The Android support library com.android.support:exifinterface was added as some required features are lacking from earlier Android versions.

As a minor detail, strings.xml was standardized to 4 spaces per tab. That's why there are so many removed lines.

## Tests performed (required)

Tested on API 22 Genymotion Google Nexus 5 emulator, with build variant, BetaDebug and ProdDebug.

## Screenshots showing what changed (optional)
 
![image](https://user-images.githubusercontent.com/7621539/42709647-d424fe44-86e9-11e8-95bd-6de48f231621.png)
![image](https://user-images.githubusercontent.com/7621539/42709669-df1740dc-86e9-11e8-89f9-a3fdd435d7d6.png)
![image](https://user-images.githubusercontent.com/7621539/42709675-e807321a-86e9-11e8-897b-10d78588011c.png)

ilgazer commented:
Maybe the next step would be adding options for setting EXIF and location behaviour per image in a way that would also be a permanent solution to #1599 Update "automatically get current location" setting text to inform that files will be geotagged with it

ilgazer commented:
I'm not sure why the conflicts remain as there seems to be nothing left that needs changing to allow a merge.

neslihanturan commented:
Hi @ilgazer thanks for your PR but it is huge, which is something makes tester unhappy:/ 
- Can you please revert strings.xml file commit, code quality improvements should be in a separate PR
- Can you verify that all changes here are related with problem you solved? If so I will start to dig in.
- preferences.xml is changed recently. It would be great if you can solve conflicts.
Thanks in advance!

ilgazer commented:
Hey @neslihanturan, I know the PR is a bit big but it's all related to
#1686. Also, I have fixed the strings.xml issue. I also tried to solve the
conflicts but the conflict markers seem to still be there. I will look into
it some more when I have some time.

On 22 Jul 2018 18:01, "neslihanturan" <notifications@github.com> wrote:

Hi @ilgazer <https://github.com/ilgazer> thanks for your PR but it is huge,
which is something makes tester unhappy:/

   - Can you please revert strings.xml file commit, code quality
   improvements should be in a separate PR
   - Can you verify that all changes here are related with problem you
   solved? If so I will start to dig in.
   - preferences.xml is changed recently. It would be great if you can
   solve conflicts.
   Thanks in advance!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/commons-app/apps-android-commons/pull/1712#issuecomment-406872950>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AHRLo8ceVUKm4AWqLUUXq4m8ahcJ_jeCks5uJJPJgaJpZM4VPYZ5>
.

neslihanturan commented:
@ilgazer , sorry for late testing, I was busy with other tasks in project. Currently I tried to remove Camera Model EXIF and uploaded a photo. But EXIF was still there, here is the photo I uploaded after selecting Remove EXIF tag checkbox: https://commons.m.wikimedia.org/wiki/File:Triangular_bungalow_on_seaside.jpg

Besides functional problem, I think removing something by selecting check boxes doesn't feel very natural from user interface point of view. I think, instead of **Remove EXIF Tags**  item, we can have **Manage EXIF Tags** item, and behind, _Dislplay author_, _Display Copyritght_ checkbox items. Selected means we will include that EXIF tag (so they will all be selected as default, user can deselect if they want), unelected means it will be hidden.

ilgazer commented:
Your UI recommendation really makes sense and would actually be easy to
implement. I will also look at the first bug tomorrow.

On 22:40, Wed, 29 Aug 2018 neslihanturan, <notifications@github.com> wrote:

> @ilgazer <https://github.com/ilgazer> , sorry for late testing, I was
> busy with other tasks in project. Currently I tried to remove Camera Model
> EXIF and uploaded a photo. But EXIF was still there, here is the photo I
> uploaded after selecting Remove EXIF tag checkbox:
> https://commons.m.wikimedia.org/wiki/File:Triangular_bungalow_on_seaside.jpg
>
> Besides functional problem, I think removing something by selecting check
> boxes doesn't feel very natural from user interface point of view. I think,
> instead of *Remove EXIF Tags* item, we can have *Manage EXIF Tags* item,
> and behind, *Dislplay author*, *Display Copyritght* checkbox items.
> Selected means we will include that EXIF tag (so they will all be selected
> as default, user can deselect if they want), unelected means it will be
> hidden.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/commons-app/apps-android-commons/pull/1712#issuecomment-417079493>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHRLo43jwqs38AUSh4-R2xdAS9neWWELks5uVu4wgaJpZM4VPYZ5>
> .
>

ilgazer commented:
Ok, so I implemented a new setting to remove XMP data, which is something similar to EXIF, but there are no good Android libraries to edit XMP data. Because of this, I only implemented deleting all XMP data by parsing the raw jpeg file and finding the XMP section. A future patch for better XMP support could be considered but that could easily turn into implementing a whole library, which is something I have inadequate time for. 

I also changed to Manage Exif tags and the ticked items being kept, which means that if there are not new bugs, this PR should finally be ready to merge.

neslihanturan commented:
It works as intented now, but since I think this PR can be a little tricky (it effects uploading) I request test from someone else too. But according to my test, it is ok.

misaochan commented:
Thanks for the PR, @ilgazer ! I think the code is very well-structured, and your Javadocs are excellent. :)

I am very sorry that I missed the conversation at #181 - I personally disagree with reducing accuracy by 20km or even 10km. IMO, if the user is very concerned about anonymity, they should be encouraged to choose "Remove all location data", not to submit data that is up to 20km off. Because, I think, it is better to submit a picture with no geotag to Commons than one with such a markedly wrong geotag. Also, it seems that @jayvdb 's suggestion in that issue was to REMOVE the geotag if it fell within a certain range of home, not to modify it to a different geotag.

However, switching to the "remove if near home" method would be very time-consuming and this PR is very complete. And, 100m or even 1km would be within the usual error range of GPS so the geotag might still be useful. So I would propose removing the 5km, 10km and 20km options from the Setting, and only letting the user choose 100m or 1km. @neslihanturan @nicolas-raoul @VojtechDostal What do you think?

misaochan commented on app/src/main/java/fr/free/nrw/commons/ui/LongTitlePreferences/LongTitleMultiSelectListPreference.java:
```diff
@@ -0,0 +1,30 @@
+import android.widget.TextView;
+
+/**
+ * Created by Ilgaz Er on 7/31/2018.
```
Please remove author tags, I know Android Studio automatically adds them, but our code style discourages them. :)

ilgazer commented:
Hmm, fair enough about introducing large errors to GPS coordinates. By the
way, disabling the GPS feature wouldn't really take much time, not more
than 5 minutes. :) Maybe we should approve this PR without the Imprecise
GPS feature and ask the commons community. After all, we can always
retrieve the code from Git history if the COmmons people don't object.

On Sun, 9 Sep 2018 at 18:20, VojtechDostal <notifications@github.com> wrote:

> I think we should be really careful with submitting deliberately imprecise
> coordinates. What does the Wikimedia Commons community think about that?
> Even 100m could be controversial. Some tools display Commons images as pins
> on a map. 100m shift will place the pins to a completely different position
> on a map. I'd rather add no coordinates than imprecise coordinates (or, if
> we decide to submit imprecise coordinates, we should be explicit about that
> and warn potential users using a tag or a template parameter).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/commons-app/apps-android-commons/pull/1712#issuecomment-419723060>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHRLo3NDSKXCO4JsIpl6c3r8yybelSS6ks5uZTGjgaJpZM4VPYZ5>
> .
>

VojtechDostal commented:
Then we should be more open about that and specify the precision used to fill in the template.

misaochan commented:
> Maybe we should approve this PR without the Imprecise GPS feature and ask the commons community. After all, we can always retrieve the code from Git history if the Commons people don't object.

@ilgazer Sure, that sounds like a good plan, let's go with that. :) Let us discuss this further at #181 , and it would be great if we could ask in the Commons Village Pump as well. In the meantime, you could remove the GPS accuracy Setting and fix the other bugs, so we can re-test and merge this PR while discussion is ongoing? It is better to have more modular PRs anyway. ;)

misaochan commented:
Just to clarify, the remaining bugs (which I added in an edit later, so it might not be visible to those using email updates) are:

I have tested the PR on API 24 Nexus S emulator. My uploads work as usual, but:

- Please set a default value (Do Nothing) in settings, otherwise it is unclear to the user what the default is. Also, display the current value chosen as subtext (similar to Recent Upload Limit).
- Tapping the Set XMP Metadata option does nothing?
- I chose "Do Nothing" in Location accuracy, and uploaded a picture. The picture in Commons has no geotag even though in my gallery view it has a location. https://commons.wikimedia.org/wiki/File:Yong_tau_fu.jpg Tested twice and this was still the case.
- Actually, more generally, all of my EXIF is always stripped regardless of what my selection for the "Manage EXIF tags" setting is. @neslihanturan do your coords and EXIF data work with this PR?

ilgazer commented:
> Please set a default value (Do Nothing) in settings, 

Good catch! Fixed it.

>Tapping the ~Set~ Keep XMP Metadata option does nothing?

Could your image have no XMP metadata? You can upload the picture here to see its XMP metadata: http://metapicz.com

> Actually, more generally, all of my EXIF is always stripped regardless of what my selection for the "Manage EXIF tags" setting is. @neslihanturan do your coords and EXIF data work with this PR?

This seems incredibly odd. The only piece of code capable of doing that is the XMP removal code. Could you give a link to an image that is losing its EXIF data on upload?

misaochan commented:
@ilgazer https://commons.wikimedia.org/wiki/File:Yong_tau_fu.jpg is one of them.

misaochan commented:
@ilgazer Shall I email you the picture? I don't think Commons will be happy with me uploading a duplicate, lol.

ilgazer commented:
I would love that

ilgazer commented:
That would be great! :)

On 09:12, Tue, 18 Sep 2018 Josephine Lim, <notifications@github.com> wrote:

> @ilgazer <https://github.com/ilgazer> Shall I email you the picture? I
> don't think Commons will be happy with me uploading a duplicate, lol.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/commons-app/apps-android-commons/pull/1712#issuecomment-422267773>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHRLo2qwNT6K2Mac2Bd5HBLfwTOKO0FBks5ucI7ggaJpZM4VPYZ5>
> .
>

misaochan commented:
IIRC I emailed @ilgazer the image, but he hasn't gotten back to me about it?

neslihanturan commented:
Well, I am not sure what to do with this PR, seems like it is sleeping for a while. It has conflicts and known issues. On the other hand, would be great to have this code after it is fixed. So my vote is keeping this open, someone else can take this up and fix since the code looks good overall.

ilgazer commented:
I will be terribly busy till the middle of June, so I agree that that
someone else taking over the PR would be the best course of action.

VitalyVPinchuk commented:
Hello!
Need some help.
I'm trying to merge changes from ilgazer:master to up-to-date master and having some troubles.
For example, FileProcessor class uses FileUtils.copy() method, which is no longer available.
Should I put copy() back into FileUtils class or is there a better solution?

ilgazer commented:
No you shouldn't. The purpose all that function was to make sure that the
file was copied before we destroy its metadata. Honestly I think you should
rethink where the metadata removal should be done and tie in the EXIF
clensing code accordingly.

On Mar 23, 2019 3:21 PM, "Vitaly V. Pinchuk" <notifications@github.com>
wrote:

> Hello!
> Need some help.
> I'm trying to merge changes from ilgazer:master to up-to-date master and
> having some troubles.
> For example, FileProcessor class uses FileUtils.copy() method, which is no
> longer available.
> Should I put copy() back into FileUtils class or is there a better
> solution?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/commons-app/apps-android-commons/pull/1712#issuecomment-475865364>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHRLo6mTh4G3JiGkRdo9KgK8kuY9wz4hks5vZhxmgaJpZM4VPYZ5>
> .
>

VitalyVPinchuk commented:
@ilgazer, I'm trying to use ExifInterface and it's not clear for me which one I should prefer:
(1) android.media.ExifInterface
(2) android.support.media.ExifInterface
(3) androidx.exifinterface.media.ExifInterface

By the way, it seems that only (1) returns image's latitude and longitude with getLatLong(float[]). Methods getLatLong() of both (2) and (3) return null despite the existence of image's location tag.
What am I doing wrong?

VitalyVPinchuk commented:
Please have a look at these two images. The smaller one was taken in emulator and the other one with my OP6. Both of them are shown as having geo tags in Google Photos app. But only OP6's image seems to have lat/long tags while reading with android.support.media.ExifInterface.getLatLong().
Why is it so?

![IMG_20190329_144354](https://user-images.githubusercontent.com/26341194/55236555-4413a200-5241-11e9-8b12-fe25a9305e33.jpg)
![IMG_20190329_162205](https://user-images.githubusercontent.com/26341194/55236556-44ac3880-5241-11e9-8926-18896f7679fc.jpg)

ilgazer commented:
I don't remember which one works but you should use the one I used
previously. I am not sure if I used the other one anywhere. If I used both,
use the one that works.
